samcli/commands/publish/command.py

```diff
@@ -122,14 +133,19 @@ def _wrap_s3_uri_exception(ex):
     Returns
     -------
     Exception
-        UserException if found invalid S3 URI or ClientError
+        UserException if found invalid S3 URI or SemVer, else ClientError
     """
     error_code = ex.response.get('Error').get('Code')
     message = ex.response.get('Error').get('Message')
 
-    if error_code == 'BadRequestException' and "Invalid S3 URI" in message:
-        return UserException(
-            "Your SAM template contains invalid S3 URIs. Please make sure that you have uploaded application "
-            "artifacts to S3 by packaging the template: 'sam package --template-file <file-path>'.")
+    if error_code == 'BadRequestException':
```

Comment:
I'm not too fond of looking at the message of the exception to determine the exception we throw, I dont know if there is a better way to do this though.

Reply:
The only way to tell whether this exception is about S3 or Version is from the message, because we throw general BadRequestException 😞: https://docs.aws.amazon.com/serverlessrepo/latest/devguide/applications-applicationid-versions-semanticversion.html.

Reply:
Are these the only ones we need to be catching? What is thrown when the semver already exists? Is throttle a concern?

To @TheSriram note, can this be handled within the `publish_application` call? Instead of consumers of the library having to deal with all these cases, you could expand the library to throw more friendly exceptions. This would result in cleaner code here and for any other consumers.

Reply:
Makes sense, I'll handle the exceptions inside serverlessrepo, except the S3 URI one. I still want to keep it here since the error message would be specific to sam cli (telling people to run sam package), what do you think?